paths_cli/compiling/core.py

```diff
@@ -0,0 +1,300 @@
+        primary callable to build an object, or string representing the
+        fully-qualified path to a callable
+    remapper : Callable[[Dict], Dict], optional
+        callable to remap the the mapping of ???
+    after_build : Callable[[Any, Dict], Any], optional
+        ccallable to update the created object with any additional
+        information from the original dictionary.
+    """
+    def __init__(self, builder, *, remapper=None, after_build=None):
+        if remapper is None:
+            remapper = lambda dct: dct
+        if after_build is None:
+            after_build = lambda obj, dct: obj
+        self.remapper = remapper
+        self.after_build = after_build
+        self.builder = builder
+
+    def __call__(self, **dct):
+        # TODO: change this InstanceCompilerPlugin.build to make this better
+        if isinstance(self.builder, str):
+            module, _, func = self.builder.rpartition('.')
+            builder = import_thing(module, func)
+        else:
+            builder = self.builder
+
+        ops_dct = self.remapper(dct.copy())
+        obj = builder(**ops_dct)
+        after = self.after_build(obj, dct.copy())
+        return after
+
+
+class InstanceCompilerPlugin(OPSPlugin):
+    """
+
+    Parameters
+    ----------
+    builder : Callable
+        Function that actually creates an instance of this object. Note that
+        the :class:`.Builder` class is often a useful tool here (but any
+        callable is allowed).
+    parameters : List[:class:`.Parameter]
+        Descriptions of the paramters for that the builder takes.
+    name : str
+        Name used in the text input for this object.
+    aliases : List[str]
+        Other names that can be used.
+    requires_ops : Tuple[int, int]
+        version of OpenPathSampling required for this functionality
+    requires_cli : Tuple[int, int]
+        version of the OPS CLI requires for this functionality
+    """
+    SCHEMA = "http://openpathsampling.org/schemas/sim-setup/draft01.json"
+    category = None
+    def __init__(self, builder, parameters, name=None, aliases=None,
+                 requires_ops=(1, 0), requires_cli=(0, 3)):
+        super().__init__(requires_ops, requires_cli)
+        self.aliases = aliases
+        self.attribute_table = {p.name: p.loader for p in parameters
+                                if p.required}
+        self.optional_attributes = {p.name: p.loader for p in parameters
+                                    if not p.required}
+        self.defaults = {p.name: p.default for p in parameters
+                         if not p.required}
+        self.name = name
+        self.builder = builder
+        self.builder_name = str(self.builder)
+        self.parameters = parameters
+        logger_name = f"compiler.InstanceCompilerPlugin.{builder}"
+        self.logger = logging.getLogger(logger_name)
+
+    @property
+    def schema_name(self):
+        if not self.name.endswith(self.category):
```

Comment:
this can error out is `category` is still `None` when this property is called

Reply:
updated to

```python
if self.category and not self.name.endswith(self.category):
```